### PR TITLE
New location validity check

### PR DESCRIPTION
**Pull request by michalisFr**
Added a validity check for location:
- New `checkLocation` function (with hard-coded sanctioned locations) in `ValidityChecks.ts`
- New `setLocationValidity` function in `Candidate.ts`
- New `locationValid` check in `CheckCandidates.ts`
- Updates to `InvalidityReasonType` and `InvalidityReason` in `models.ts`

### CODE REVIEW

File: packages/common/src/constraints/CheckCandidates.ts [hunk 1]
```diff
@@ -142,7 +149,8 @@ export const checkCandidate = async (
       blockedValid &&
       kusamaValid &&
       providerValid &&
-      beefyValid;
+      beefyValid &&
+      locationValid;
```
Comment (ironoa) [inline packages/common/src/constraints/CheckCandidates.ts:153]: location as a name is misleading in this context: location unfortunately encompass more than just country and region (the things that you are interested to filter), i.e. provider, cpu, ... 

thus the fact that you are checking locationValid is not true (i.e. provider is checked 2 lines above)

I suggest a name such as `geoAreaValid` or similar, and if this is targeting just the sanctions I'd go even further with `geoAreaSanctionsValid`

File: packages/common/src/db/queries/Candidate.ts [hunk 2]
```diff
@@ -1469,6 +1469,52 @@ export const setProviderInvalidity = async (
...
+              ? ""
+              : details
+                ? details
+                : `${data.name} is in a sanctioned location`,
+          },
+        ],
+      },
+    ).exec();
+  } catch (e) {
+    logger.info(`error setting provider validity`);
```
Comment (ironoa) [inline packages/common/src/db/queries/Candidate.ts:1514]: ? Provider ?

File: packages/common/src/db/models.ts [hunk 3]
```diff
@@ -88,6 +88,7 @@ export enum InvalidityReasonType {
   KUSAMA_RANK = "KUSAMA_RANK",
   PROVIDER = "PROVIDER",
   BEEFY = "BEEFY",
+  SANCTIONS = "SANCTIONS",
```
Comment (ironoa) [inline packages/common/src/db/models.ts:91]: SANCTIONED_GEO_AREA

**Review by ironoa — CHANGES_REQUESTED** [review]

File: packages/common/src/constraints/ValidityChecks.ts [hunk 4]
```diff
@@ -390,6 +391,47 @@ export const checkProvider = async (
...
+): Promise<boolean> => {
+  try {
+    const location = await queries.getCandidateLocation(
+      candidate.slotId,
+      candidate.name,
+      candidate.stash,
+    );
+    if (location && location.region && location.country) {
+      // const sanctionedLocations = config.telemetry?.sanctionedLocations;
+      const sanctionedCountries = ["RU", "IR", "CU", "KP", "SY"];
```
[the remark this answers is not in the document]
Comment (michalisFr) [reply packages/common/src/constraints/ValidityChecks.ts:408]: @ironoa The config files are under `packages > core > config` correct? In which ones should these arrays be added? All of them?

[the remark this answers is not in the document]
Comment (ironoa) [reply packages/common/src/constraints/ValidityChecks.ts:408]: I'd add them in the [constraints](https://github.com/w3f/1k-validators-be/blob/staging/packages/core/config/kusama.current.sample.json#L37) section, subsection sanctionedCountries and sanctionedRegion

[the remark this answers is not in the document]
Comment (michalisFr) [reply packages/common/src/constraints/ValidityChecks.ts:408]: > I'd add them in the [constraints](https://github.com/w3f/1k-validators-be/blob/staging/packages/core/config/kusama.current.sample.json#L37) section, subsection sanctionedCountries and sanctionedRegion

Ok, I was going to add them in the telemetry section, but that works too. In which config files though? There are 9 config files in that folder. Add it to all of them?

[the remark this answers is not in the document]
Comment (ironoa) [reply packages/common/src/constraints/ValidityChecks.ts:408]: all of them (they are mostly samples that have to be aligned for coherence). Don't forget the helm [value file](https://github.com/w3f/1k-validators-be/blob/master/charts/otv-backend/values.yaml)

**Comment by michalisFr** [comment]
> this PR should be targeted to staging, not to master

Done

File: packages/common/src/constraints/ValidityChecks.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (michalisFr) [reply packages/common/src/constraints/ValidityChecks.ts]: > all of them (they are mostly samples that have to be aligned for coherence). Don't forget the helm [value file](https://github.com/w3f/1k-validators-be/blob/master/charts/otv-backend/values.yaml)

Done. I also added two more missing values

File: candidates/kusama.json [hunk 5]
```diff
@@ -3119,8 +3119,8 @@
     },
     {
```
Comment (ironoa) [inline candidates/kusama.json:3120]: ?

File: candidates/polkadot.json [hunk 6]
```diff
@@ -2287,6 +2287,22 @@
       "kusamaStash": "FkGx1DCPF85BRjGZszZCXL7GSoWKHASy3jP7iuWV9pbHQVq",
       "riotHandle": "@colossus.digital:matrix.org",
       "kyc": true
```
Comment (ironoa) [inline candidates/polkadot.json:2289]: ?

File: charts/otv-backend/values.yaml [hunk 7]
```diff
@@ -48,7 +48,11 @@ config: |
     },
     "constraints": {
       "skipConnectionTime": false,
-      "skipIdentity": false
+      "skipIdentity": false,
+      "skipClientUpgrade": false,
+      "skipUnclaimed": false,
```
Comment (ironoa) [inline charts/otv-backend/values.yaml:53]: ?

File: packages/common/src/constraints/CheckCandidates.ts [hunk 8]
```diff
@@ -124,6 +125,11 @@ export const checkCandidate = async (
       logger.info(`${candidate.name} provider not valid`, constraintsLabel);
     }
 
+    const geoAreaSanctionsValid =
+      constraints.config?.constraints?.skipSanctionedCountries == true
+        ? true
+        : (await checkLocation(constraints.config, candidate)) || false;
```
Comment (ironoa) [inline packages/common/src/constraints/CheckCandidates.ts:131]: checkLocation has still the old naming...

File: packages/common/src/db/queries/Candidate.ts [hunk 9]
```diff
@@ -1350,6 +1350,21 @@ export const setProviderInvalidity = async (
   }
 };
 
+// Set Sanctions Validity Status
+export const setLocationValidity = async (
```
Comment (ironoa) [inline packages/common/src/db/queries/Candidate.ts:1354]: wrong naming

File: packages/common/src/constraints/CheckCandidates.ts [hunk 10]
```diff
@@ -124,6 +125,11 @@ export const checkCandidate = async (
       logger.info(`${candidate.name} provider not valid`, constraintsLabel);
     }
 
+    const geoAreaSanctionsValid =
+      constraints.config?.constraints?.skipSanctionedCountries == true
```
Comment (ironoa) [inline packages/common/src/constraints/CheckCandidates.ts:129]: why you defined skipSanctionedCountries and skipSanctionedRegions but you are checking only Countries ? couldn't you just define a skipSanctionedArea flag ?

File: packages/core/config/polkadot.current.sample.json [hunk 11]
```diff
@@ -11,14 +11,18 @@
     "candidatesUrl": "https://raw.githubusercontent.com/w3f/1k-validators-be/master/candidates/polkadot.json"
   },
   "constraints": {
-    "skipConnectionTime": false,
+    "skipConnectionTime": true,
```
Comment (ironoa) [inline packages/core/config/polkadot.current.sample.json:14]: ??

File: packages/core/config/polkadot.microservice.dev.sample.json [hunk 12]
```diff
@@ -13,14 +13,18 @@
     "candidatesUrl": "https://raw.githubusercontent.com/w3f/1k-validators-be/master/candidates/polkadot.json"
   },
   "constraints": {
-    "skipConnectionTime": false,
+    "skipConnectionTime": true,
```
Comment (ironoa) [inline packages/core/config/polkadot.microservice.dev.sample.json:16]: ????

File: packages/core/config/polkadot.microservice.sample.json [hunk 13]
```diff
@@ -13,14 +13,18 @@
     "candidatesUrl": "https://raw.githubusercontent.com/w3f/1k-validators-be/master/candidates/polkadot.json"
   },
   "constraints": {
-    "skipConnectionTime": false,
+    "skipConnectionTime": true,
```
Comment (ironoa) [inline packages/core/config/polkadot.microservice.sample.json:16]: ??

**Review by ironoa — CHANGES_REQUESTED** [review]

**Pull request closed without merging**